### PR TITLE
SSR

### PR DESCRIPTION
**Pull request by luxaritas**
This one was a bear but finally got it all working. LMK if anything looks suspect

### CODE REVIEW

**Comment by luxaritas** [comment]
Note one thing I also did was set up vuex-class-module proxies. Type safety ftw!

File: README.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ The rewrite is based on the design present [here](https://github.com/eternagame/
 ```
 npm install
 ```
-Also, copy the `.env.example` file to `.env`, and fill in your credentials.
+If you want to modify any of the parameters in `.env`, create a file `.env.local` and include any overrides there
```
Comment (guyguy2001) [inline README.md]: I think "create a `.env.local` file" sounds better

File: server/index.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    }
+  }
+
+  // Contrary to what you'd expect, the first env loaded is the one whose values are kept
+  loadEnv(resolve('../.env.local'));
+  loadEnv(resolve('../.env'));
+}
+
+const PORT = process.env.PORT;
+const HOST = process.env.HOST;
```
Comment (guyguy2001) [inline server/index.js]: I feel like something like INSTANCE_HOST/PORT would be better, unless naming the local server's address `HOST`/`PORT` is a .env convention.

File: src/App.vue [hunk 3]
```diff
@@ -20,6 +20,9 @@
     },
   })
   export default class App extends Vue {
+    async fetchPageData() {
+      this.$vxm.user.authenticate();
+    }
```
Comment (guyguy2001) [inline src/App.vue]: Oh, so I don't even have to do it myself; Sweet!

File: src/app.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+Vue.use(BootstrapVue);
+
+InitGlobalComponents();
+
+Vue.config.productionTip = false;
+
+export default function createApp() {
+  const router = createRouter();
+  const store = createStore();
```
Comment (guyguy2001) [inline src/app.ts]: Just to make sure I understand - Is this code ran on the same server instance for multiple instances of Vue?

File: src/app.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  // Types for properties added to Vue (prototype) are in types/vue.d.ts
+  // Make sure to update that if you change what is included here!
+
+  const axiosInstance = axios.create({
+    withCredentials: true,
+  });
+  Vue.$http = axiosInstance;
+  Vue.prototype.$http = axiosInstance;
```
Comment (guyguy2001) [inline src/app.ts]: Why not `$axios`?
If it's for the sake of conventions / being generic, I'd say I'm against it, as this behaves differently from the vue-resource `$http` object, even though they serve the same purpose.

File: src/app.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+  Vue.prototype.$vxm = {
+    user: createProxy(store, UserStore),
+    mobile: createProxy(store, MobileStore),
+    pageData: createProxy(store, PageData),
+  };
```
Comment (guyguy2001) [inline src/app.ts]: I think this should be a non-default export of `store.ts`, so that we don't have to re-declare the modules on two different files.

File: src/components/PageLayout/Navbar/NavbarMenu/Login/LoginSub.vue [hunk 7]
```diff
@@ -21,12 +21,14 @@
...
+          password: this.password,
+        });
         if (data.success) {
           this.$router.push('/labs/explore');
         } else {
-          this.$store.commit('showLoginFailedModal', { errorMessage: data.error });
+          this.$vxm.user.showLoginFailedModal({ errorMessage: data.error });
         }
       }
     }
```
Comment (guyguy2001) [inline src/components/PageLayout/Navbar/NavbarMenu/Login/LoginSub.vue]: My bad on this, but this component shouldn't have anything inside of it.
I don't mind to remove it in a later commit, but if you're already editing it, you can drop it.

File: src/components/PageLayout/Navbar/NavbarMenu/NavbarCollapseContent.vue [hunk 8]
```diff
@@ -35,11 +35,8 @@
       [text: string]: {[text: string]: string} | string;
     };
 
-    loggedIn = false;
-
-    @Watch('$route')
-    checkLoginStatus() {
-      this.loggedIn = this.$store.state.userStore.loggedIn;
+    get loggedIn() {
+      return this.$vxm.user.loggedIn;
```
Comment (guyguy2001) [inline src/components/PageLayout/Navbar/NavbarMenu/NavbarCollapseContent.vue]: The reason I waited for a route change is that before SSR, `user.loggedIn` would have been set to `true` before the new page loaded, and I wanted everything to change at once.
Did that change on the SSR version?

File: src/main.ts [hunk 9]
```diff
@@ -1,21 +0,0 @@
-import Vue from 'vue';
```
Comment (guyguy2001) [inline src/main.ts]: You should rename main.ts to app.ts, instead of removing/creating them, in my opinion.

File: src/store/mobile.vuex.ts [hunk 10]
```diff
@@ -1,12 +1,9 @@
 import {
   createModule, mutation, action, extractVuexModule,
 } from 'vuex-class-component';
-import axios from 'axios';
-
-axios.defaults.withCredentials = true;
 
 const VuexModule = createModule({
-  namespaced: 'mobileStore',
+  // namespaced: 'mobileStore',
```
Comment (guyguy2001) [inline src/store/mobile.vuex.ts]: Might as well remove it

Comment (guyguy2001) [reply src/store/mobile.vuex.ts]: Is this because of the bug I've experienced with the user store? Or is it because it is no longer needed?

File: src/types/vue.d.ts [hunk 11]
```diff
@@ -14,6 +19,21 @@ declare module 'vue/types/vue' {
...
+      pageData: ProxyWatchers & PageData,
+    }
+  }
+
+  interface VueConstructor {
+    $http: AxiosInstance,
+    $vxm: {
+      user: ProxyWatchers & UserStore,
+      mobile: ProxyWatchers & MobileStore,
+      pageData: ProxyWatchers & PageData,
```
Comment (guyguy2001) [inline src/types/vue.d.ts]: I don't see another way, but I'm not a big fan of redeclaring it...
Do you see a way to change the vuex-class-component module to allow for this to happen automatically?
Or perhaps we could import this type from `store.ts`?

File: src/views/labs/LabView/LabView.vue [hunk 12]
```diff
@@ -1,5 +1,5 @@
 <template>
-  <EternaPage v-if="data" title="Lab Details">
+  <EternaPage v-if="lab" title="Lab Details">
```
Comment (guyguy2001) [inline src/views/labs/LabView/LabView.vue]: I think I prefer `data`... But I'll have to think about it more

File: src/views/labs/LabsExplore/LabsExplore.vue [hunk 13]
```diff
@@ -30,34 +32,35 @@
...
     async beforeRouteUpdate(to: Route, from: Route, next: RouteCallback<LabsExploreView>) {
-      const data = await LabsExploreView.query(to);
-      this.data = data;
+      const data = await this.query(to);
+      this.dynamicData = data;
       next();
     }
 
-    private filters = [new Filter('active', 'Active'), new Filter('inactive', 'Inactive')];
+    private filters: Filter[] = [{ value: 'active', text: 'Active' }, { value: 'inactive', text: 'Inactive' }];
```
Comment (guyguy2001) [inline src/views/labs/LabsExplore/LabsExplore.vue]: Note to self - check if this has automplete

File: server/index.js (hunk 2 above)
Comment (luxaritas) [reply server/index.js]: `HOST` and `PORT` are nodejs conventions

File: src/app.ts (hunk 4 above)
Comment (luxaritas) [reply src/app.ts]: The server will import the frontend code once, but will call createApp once for each new connection, so it will have a fresh app, router, and store each time

File: src/app.ts (hunk 5 above)
Comment (luxaritas) [reply src/app.ts]: I am doing this for the sake of convention - from my understanding, in Vue (and other libraries like Angular) whatever service you're using for networking is located at `$http`. I wouldn't be torn up if we used `$axios`, but I definitely wanted to make sure it was prefixed with `$` (rather than before where it was just `axios`)

File: src/app.ts (hunk 6 above)
Comment (luxaritas) [reply src/app.ts]: Issue is, we can't - the store instance is created within `createStore` (since we have to create fresh instances for SSR rather than creating it once when the file is first imported), so there's no way to export it. I also don't think it would be smart to export a default null variable that gets filled each time createStore is called, as I think there could be multiple instances at once when there are multiple SSR requests at the same time.

File: src/main.ts (hunk 9 above)
Comment (luxaritas) [reply src/main.ts]: Nothing I can control, that's automatically determined by Vue :(

File: src/store/mobile.vuex.ts (hunk 10 above)
Comment (luxaritas) [reply src/store/mobile.vuex.ts]: Yup same bug. I actually found the cause, but haven't deduced how to fix it yet. I'm hoping that I'll be able to...

File: src/types/vue.d.ts (hunk 11 above)
Comment (luxaritas) [reply src/types/vue.d.ts]: I don't see any other option since it's defined on the Vue prototype post-hoc, so it can't be statically inferred afaik

File: src/views/labs/LabView/LabView.vue (hunk 12 above)
Comment (luxaritas) [reply src/views/labs/LabView/LabView.vue]: We can't use `data` with a computed property - it gets clobbered (per our discussion the other day). I'm also not sure why it's particularly helpful to make the differentiation?

File: src/views/labs/LabsExplore/LabsExplore.vue (hunk 13 above)
Comment (luxaritas) [reply src/views/labs/LabsExplore/LabsExplore.vue]: Interfaces do indeed have autocomplete!

File: src/components/PageLayout/Navbar/NavbarMenu/NavbarCollapseContent.vue (hunk 8 above)
Comment (luxaritas) [reply src/components/PageLayout/Navbar/NavbarMenu/NavbarCollapseContent.vue]: The issue with that is then it won't get updated on the initial `authenticate` call, so if you're already logged in, it won't reflect that until you change pages

File: README.md (hunk 1 above)
Comment (luxaritas) [reply README.md]: How about "create a file named `.env.local`"?

File: src/components/PageLayout/Navbar/NavbarMenu/Login/LoginSub.vue (hunk 7 above)
Comment (luxaritas) [reply src/components/PageLayout/Navbar/NavbarMenu/Login/LoginSub.vue]: Done

File: README.md (hunk 1 above)
Comment (guyguy2001) [reply README.md]: Great

File: server/index.js (hunk 2 above)
Comment (guyguy2001) [reply server/index.js]: 👍

File: src/app.ts (hunk 6 above)
Comment (guyguy2001) [reply src/app.ts]: I see...
Then maybe we can return it in the return value of createStore (by making it return an object of the store and the proxies)?
That would be weird to use though...

File: src/components/PageLayout/Navbar/NavbarMenu/NavbarCollapseContent.vue (hunk 8 above)
Comment (guyguy2001) [reply src/components/PageLayout/Navbar/NavbarMenu/NavbarCollapseContent.vue]: How about `@Watch('$route', { immediate: true })`?

File: src/app.ts (hunk 6 above)
Comment (luxaritas) [reply src/app.ts]: It would still need to be added to Vue.prototype in able to be accessible by all components...

File: src/components/PageLayout/Navbar/NavbarMenu/NavbarCollapseContent.vue (hunk 8 above)
Comment (guyguy2001) [reply src/components/PageLayout/Navbar/NavbarMenu/NavbarCollapseContent.vue]: I want the page to only load after the authentication is done, so that the login menu doesn't disappear after it was loaded; I'll have to check whether that's how it is now, but it wouldn't make sense otherwise in my opinion.

Comment (luxaritas) [reply src/components/PageLayout/Navbar/NavbarMenu/NavbarCollapseContent.vue]: That will still get fired off before the authenticate request has completed won't it?

File: src/views/labs/LabView/LabView.vue (hunk 12 above)
Comment (guyguy2001) [reply src/views/labs/LabView/LabView.vue]: My thought is that it's part of the template for pages with a query - only show the content if your query is done, regardless of its content.

File: src/main.ts (hunk 9 above)
Comment (guyguy2001) [reply src/main.ts]: Can't you use `git mv` for that?

Comment (guyguy2001) [reply src/main.ts]: Afaik you can still use it after the original file is removed

File: src/app.ts (hunk 5 above)
Comment (guyguy2001) [reply src/app.ts]: From the vue-axios examples page:
```
this.axios.get(api).then((response) => {
  console.log(response.data)
})

this.$http.get(api).then((response) => {
  console.log(response.data)
})
```
Looks like I misremembered how it was there; So it seems like it is a bigger convention than I thought.

File: src/app.ts (hunk 6 above)
Comment (guyguy2001) [reply src/app.ts]: When we would call createStore, we would save it into `store` and `proxies` variables, and then pass each of them to the right place.

File: src/mixins/PageData.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+ * Abstraction for handling fetching data for a page while allowing it to be cached
+ * if provided by SSR
+ *
+ * Here be abstract class mixin hackery
+ * https://github.com/vuejs/vue-class-component/issues/91#issuecomment-451422547
+ */
+// @ts-ignore
+@Component
+abstract class PageDataMixinAbstract extends Vue {
+  abstract async fetchPageData(): Promise<any>;
```
Comment (guyguy2001) [inline src/mixins/PageData.ts]: What good does it do to mark this as abstract if you derive from it immediately? Doesn't that revoke the `abstract` value?
(If not, wouldn't that break having to mark the class as `abstract` to have abstract functions?)

Comment (luxaritas) [reply src/mixins/PageData.ts]: You would think. :P Hence the "Here be abstract class mixin hackery" - read about it here: https://github.com/vuejs/vue-class-component/issues/91#issuecomment-451422547

File: src/mixins/PageData.ts [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  private async fetchPageDataInternal() {
+    this.$vxm.pageData.data = await this.fetchPageData();
+  }
+
+  async serverPrefetch() {
+    await this.fetchPageDataInternal();
+  }
+
+  mounted() {
+    if (this.$vxm.pageData.refetch || this.$vxm.pageData.data == null) {
```
Comment (guyguy2001) [inline src/mixins/PageData.ts]: Any specific reasons for not using `===`?
If not, I'm adding that as a rule (And we can always use `!(this.$vxm.pageData.data)`)

File: src/views/labs/LabView/LabView.vue (hunk 12 above)
Comment (luxaritas) [reply src/views/labs/LabView/LabView.vue]: If so, it would be `pageData`, but I don't know whether that level of consistency is really necessary... in the future we'll ideally be using graphql anyways so you won't have one consistent high-level property of the same name since you define the things you're requesting directly on the instance

File: src/mixins/PageData.ts (hunk 15 above)
Comment (luxaritas) [reply src/mixins/PageData.ts]: No, I'm surprised the linter didn't catch that (there might be some specific exclusion for null due to the way it works?)

File: src/app.ts (hunk 6 above)
Comment (luxaritas) [reply src/app.ts]: I mean all that gains us is defining it in `store.ts`... which I guess is good?

File: src/mixins/PageData.ts (hunk 14 above)
Comment (guyguy2001) [reply src/mixins/PageData.ts]: OH - I didn't realize the @ts-ignore did that.
I'll look into it, hopefully it's possible to fix it via changes to TS that we can issue (assuming they didn't already)

File: src/views/labs/LabView/LabView.vue (hunk 12 above)
Comment (guyguy2001) [reply src/views/labs/LabView/LabView.vue]: I'll see as I write more pages. At worst I'll revert it to pageData, but for now I'll keep your changes.

File: src/App.vue (hunk 3 above)
Comment (guyguy2001) [inline src/App.vue]: This doesn't call it for me - why would `fetchPageData` would be called on `App`?

File: src/views/labs/LabView/LabView.vue [hunk 16]
```diff
@@ -31,18 +31,13 @@
...
     get lab() {
-      return this.data!.lab;
+      return (this.pageData as LabViewData)?.lab;
     }
 
-    async beforeRouteEnter(to: Route, from: Route, next: RouteCallback<LabView>) {
-      const data = (await axios.get(`/get/?type=project&nid=${to.params.nid}`)).data.data as LabViewData;
-      next((vm) => {
-        vm.data = data;
-      });
```
Comment (guyguy2001) [inline src/views/labs/LabView/LabView.vue]: Why did you remove it?
I think adding it back in the mixin would be fine though
